Refactor Spec2Vec and allow both Spectrum and SpectrumDocument input

florian-huber opened the pull request:
Address #51 to make the API easier for users, and prevent issues with different decimal rounding.

- `Spec2Vec` similarity methods now work with `Spectrum` and `SpectrumDocument` input
- code examples were updated (and fixed)

florian-huber committed:
update typing

florian-huber committed:
fix typing

florian-huber committed:
refactor Spec2Vec and allow both Spectrum and SpectrumDocument input

florian-huber committed:
update code example

florian-huber committed:
fix code example

florian-huber committed:
further update and fix code examples

florian-huber committed:
further update and fix code examples 2

florian-huber committed:
Merge branch 'master' into score_class_refactoring

florian-huber committed:
update changelog

sverhoeven commented on CHANGELOG.md:
```diff
@@ -7,6 +7,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+## Changed
+
+- refactored `Spec2Vec` to now accept both `Spectrum` and `SpectrumDocument` input [#51](https://github.com/iomega/spec2vec/issues/51)
```
Sounds like you need to give Spec2Vec both Spectrum and SpectrumDocument at the same time while either one is required.
```suggestion
- refactored `Spec2Vec` to now accept `Spectrum` or `SpectrumDocument` as input [#51](https://github.com/iomega/spec2vec/issues/51)
```

sverhoeven commented on spec2vec/Spec2Vec.py:
```diff
@@ -69,43 +100,44 @@ def __init__(self, model: Word2Vec, intensity_weighting_power: Union[float, int]
         return cosine_similarity(reference_vector, query_vector)
 
-    def matrix(self, references: List[SpectrumDocument], queries: List[SpectrumDocument],
+    def matrix(self, references: Union[List[SpectrumDocument], List[SpectrumType]],
```
Probably `List[Union[SpectrumDocument, SpectrumType]]`could also work, but the type now is easier to work with as  all list members are expected to have the same type. So what you have is more future proof.

florian-huber replied:
I also briefly thought about that. But I was afraid that `List[Union[A, B]]` would suggest that could can also have a list with mixed types. In the current implementation that would even work, but it doesn't seem recommended use to me .

sverhoeven commented on spec2vec/Spec2Vec.py:
```diff
@@ -69,43 +100,44 @@ def __init__(self, model: Word2Vec, intensity_weighting_power: Union[float, int]
         ----------
         references:
-            Reference spectrum documents.
+            Reference spectrums or spectrum documents
```
Use consistent punctuation
```suggestion
            Reference spectrums or spectrum documents.
```

sverhoeven commented on spec2vec/Spec2Vec.py:
```diff
@@ -132,11 +162,26 @@ def matrix(self, references: List[SpectrumDocument], queries: List[SpectrumDocum
+        """Generate Spec2Vec embedding vectors from input spectrum (or SpectrumDocument)"""
+        if isinstance(spectrum_in, Spectrum):
+            spectrum_in = SpectrumDocument(spectrum_in, n_decimals=self.n_decimals)
+        assert spectrum_in.n_decimals == self.n_decimals, \
+            "Decimal rounding of input data does not agree with model vocabulary."
```
This can only be False when spectrum_in is already a SpectrumDocument. So assert could be skipped when spectrum_in is a Spectrum
```suggestion
        else:
            assert spectrum_in.n_decimals == self.n_decimals, \
                "Decimal rounding of input data does not agree with model vocabulary."
```

florian-huber replied:
Good point. Thanks.

sverhoeven commented on spec2vec/Spec2Vec.py:
```diff
@@ -20,31 +23,59 @@ class Spec2Vec(BaseSimilarity):
 
-        # Import pre-trained word2vec model (alternative: train new model)
-        model_file = "path and filename"
+        spectrums_file = os.path.join(os.getcwd(), "..", "tests", "pesticides.mgf")
```
It will be hard to find this file reading this on https://spec2vec.readthedocs.io/en/latest/. Can you add a download link to the example intro text with 
```rst
:download:`../tests/pesticides.mgf </../tests/pesticides.mgf>`
:download:`../integration-tests/test_user_workflow_spec2vec.model </../integration-tests/test_user_workflow_spec2vec.model>`
```
So people dont have to go GitHub to find the file.

florian-huber replied:
I will try adding this.

florian-huber replied:
Added this to the docstring. But then also realized I have no admin rights on readthedocs so I couldn't inspect the results.

sverhoeven replied:
You can build the docs locally with `cd readthedocs;make html;google-chrome _build/html/index.html`.

Added you as maintainer of https://readthedocs.org/dashboard/spec2vec/users/

florian-huber replied:
Perfect, thanks Stefan.
Don't know why I didn't think of rendering it locally.... but that of course works nicely :)
(and thanks for adding me as admin).

sverhoeven commented on spec2vec/Spec2Vec.py:
```diff
@@ -20,31 +23,59 @@ class Spec2Vec(BaseSimilarity):
+
+    .. testoutput::
+
+        Removed adduct M-H from compound name.
```
The test output is a bit minimal. Could we print the spectrum0_top10? Something like

```python
>>> [s[0].metadata['spectrumid'] for s in spectrum0_top10]
['CCMSLIB00001058300',
 'CCMSLIB00001058289',
 'CCMSLIB00001058303',
 'CCMSLIB00001058433',
 'CCMSLIB00001058430',
 'CCMSLIB00001058313',
 'CCMSLIB00001058426',
 'CCMSLIB00001058394',
 'CCMSLIB00001058298',
 'CCMSLIB00001058317']

```

```

florian-huber replied:
That's much nicer indeed. Added this.

sverhoeven reviewed (changes requested):
Nice to see improved compatibility with matchms. You retained the high code coverage and documentation renders nicely.

SpectrumType is defined as Optional[Spectrum], so it can be None, how does your code deal with None cases? Can you add tests for when spectrum = None?

I got `AssertionError: Could not find known_adduct_conversions.csv.
` error when I tried to run the Spec2Vec example code because the matchms 0.8.0 from https://anaconda.org/bioconda/matchms/ is missing the data files. It seems we have matchms on bioconda and nlesc channel. And when I did `conda env create --file conda/environment-dev.yml` the matchms from bioconda was installed. After I uninstalled/installed with pip I could run the example.

```
conda list |grep matchms
matchms                   0.8.0              pyh3252c3a_0    bioconda
```

florian-huber committed:
Apply suggestions from code review

Co-authored-by: Stefan Verhoeven <s.verhoeven@esciencecenter.nl>

florian-huber committed:
change input type to Spectrum and raise exception if wrong

florian-huber committed:
expand docstring and code example

florian-huber committed:
test name linting

florian-huber commented:
> SpectrumType is defined as Optional[Spectrum], so it can be None, how does your code deal with None cases? Can you add tests for when spectrum = None?

Thanks for spotting this Stefan! I think `None` as input should not be allowed, so I changed the input type from `SpectrumType` to `Spectrum` and added that it will raise an `ValueError` if the input type is not `Spectrum` or `SpectrumDocument`.

florian-huber committed:
try fixing docstring code example test

florian-huber commented:
> I got AssertionError: Could not find known_adduct_conversions.csv. error when I tried to run the Spec2Vec example code because ...

I added the package_data to the pypi `.tar.gz` and `matchms 0.8.1` now works (the bioconda part still misses rdkit, but that should better be dealt with in https://github.com/matchms/matchms/issues/178)

sverhoeven reviewed (approved):
Thanks for the None exception+test. The None makes sense when calling filtering functions, but not for calculating scores.

florian-huber commented:
Thanks a lot for the review Stefan!
